src/transformers/models/idefics/modeling_idefics.py

```diff
@@ -1118,14 +1148,22 @@ def forward(
             # Get sequence from the vision encoder
             image_hidden_states = self.vision_model(pixel_values=pixel_values).last_hidden_state
 
-        elif image_embeddings is not None:
-            batch_size, num_images, image_seq_len, image_hidden_size = image_embeddings.size()
-            image_hidden_states = image_embeddings.to(dtype=self.dtype, device=input_ids.device)
+        elif image_encoder_embeddings is not None:
+            batch_size, num_images, image_seq_len, image_hidden_size = image_encoder_embeddings.size()
+            image_hidden_states = image_encoder_embeddings.to(dtype=self.dtype, device=input_ids.device)
             image_hidden_states = image_hidden_states.view(batch_size * num_images, image_seq_len, image_hidden_size)
 
-        if self.config.use_resampler:
+        if self.config.use_resampler and perceiver_embeddings is None:
             image_hidden_states = self.perceiver_resampler(image_hidden_states)
-        image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+            image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+        elif self.config.use_resampler and perceiver_embeddings is not None:
+            batch_size, num_images, image_seq_len, image_hidden_size = perceiver_embeddings.size()
+            image_hidden_states = perceiver_embeddings
+        elif not self.config.use_resampler and perceiver_embeddings is None:
+            image_seq_len, image_hidden_size = image_hidden_states.size(1), image_hidden_states.size(2)
+        elif not self.config.use_resampler and perceiver_embeddings is not None:
+            raise ValueError("If perceiver_embeddings are passed, use_resampler should be True")
+
```

Comment:
for a more readable flow, i would use nested ifs:
```
if self.config.use_resampler:
    if perceiver_embeddings is None:
        do_thing
else:
    if ...
```
otherwise, the part of the logic looks good to me